### PR TITLE
bin: allow testing subsystems

### PR DESCRIPTION
**Pull request by gdams**
this PR adds the functionality to run `citgm-all` on specific tags in
the lookup by passing in the `includeTags` or `excludeTags` option.
fixes https://github.com/nodejs/citgm/issues/72

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -120,6 +122,7 @@ For syntax, see [lookup.json](./lib/lookup.json), the available attributes are:
 "sha": "<git-commit-sha>"    Test against a specific commit
 "envVar"                     Pass an environment variable before running
 "install": ["--param1", "--param2"] - Array of extra command line parameters passed to 'npm install'
+"tag": ["tag1", "tag2"]      Provide tags so that `citgm-all` can be run on specific tags
```
Comment (gibfahn) [inline README.md]: Maybe something like:

`Provide tags so that `citgm-all` can be run on specific tags` -> `Specify which tags the module comes under`

File: README.md [hunk 2]
```diff
@@ -95,6 +95,8 @@ Options:
   -a, --append                Turns on append results to file mode rather than replace
   -j, --parallel <number>     Run tests in parallel
   -J, --autoParallel          Run tests in parallel (automatically detect core count)
+  --withTags <tag>            Specify which tags to run from the lookup
+  --excludeTags <tag>         Specify which tags to skip from the lookup
```
Comment (gibfahn) [inline README.md]: Maybe:
`Specify which tags to run from the lookup` -> `Only test modules from the lookup that contain a matching tag field`

Also we should allow running against multiple tags, maybe `<tag1> <tag2>`?

File: bin/citgm-all.js [hunk 3]
```diff
@@ -93,6 +103,17 @@ function runCitgm (mod, name, next) {
     return next();
   }
 
+  if ((app.excludeTags && mod.tag && mod.tag.includes(app.excludeTags)) ||
```
Comment (MylesBorins) [inline bin/citgm-all.js]: I think it might make sense to break this out into it's own module / function that can be tested. I'm also not very thrilled about the massive boolean logic, is hard to follow and a bit fragile

File: bin/citgm-all.js [hunk 4]
```diff
@@ -93,6 +103,17 @@ function runCitgm (mod, name, next) {
...
+      (app.withTags &&  mod.tag && !mod.tag.includes(app.withTags)) ||
+        (app.withTags && !mod.tag)) {
+    log.info(name, 'no matching tag found');
+    return next();
+  }
+
+  if (app.withTags && mod.tag && mod.tag.includes(app.withTags)) {
+    log.info(name, 'the following tag was matched: ' + app.withTags);
```
Comment (MylesBorins) [inline bin/citgm-all.js]: this should include the module name

**Comment by gdams** [comment]
@MylesBorins 
we already have tests for include/exclude: https://github.com/gdams/citgm/blob/f87bb014327be7322d4c5efe5a22a2e5966b5bc3/test/bin/test-citgm-all.js#L119-L139

File: bin/citgm-all.js (hunk 4 above)
Comment (gdams) [inline bin/citgm-all.js]: This already does include the module name

**Comment by gdams** [comment]
CI: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-continuous-integration/13/

**Comment by gdams** [comment]
@nodejs-citgm PTAL

File: bin/citgm-all.js [hunk 5]
```diff
@@ -30,6 +31,14 @@ yargs = commonArgs(yargs)
     alias: 'J',
     type: 'boolean',
     description: 'Auto detect number of cores to use to run tests in parallel'
+  })
+  .option('withTags', {
+    type: 'string',
+    description: 'define which tags from the lookup to run'
+  })
+  .option('exludeTags', {
```
Comment (gibfahn) [inline bin/citgm-all.js]: Spelling

File: README.md [hunk 6]
```diff
@@ -107,6 +107,8 @@ Options:
   -j, --parallel <number>     Run tests in parallel
   -J, --autoParallel          Run tests in parallel (automatically detect core count)
   --tmpDir <path>             Directory to test modules in
+  --withTags <tag>            Only test modules from the lookup that contain a matching tag field
+  --excludeTags <tag>         Specify which tags to skip from the lookup
```
Comment (gibfahn) [inline README.md]: How do you specify multiple tags to run or exclude? Could you update the README and man pages to cover that?

Also there are no tests for multiple tags.

Comment (gibfahn) [inline README.md]: Just realised that Myles said this earlier in https://github.com/nodejs/citgm/pull/308#pullrequestreview-17039480. 

>we already have tests for include/exclude: https://github.com/gdams/citgm/blob/f87bb014327be7322d4c5efe5a22a2e5966b5bc3/test/bin/test-citgm-all.js#L119-L139

That doesn't cover multiple tags.

**Comment by gdams** [comment]
@gibfahn @MylesBorins I have now updated so that you can pass multiple tags using `citgm --withTags tag1,tag2,tag3` and have also included some tests

File: lib/lookup.js [hunk 7]
```diff
@@ -105,6 +105,9 @@ function resolve(context, next) {
             rep.script);
         context.module.script = rep.script;
       }
+      if (rep.tag) {
+        context.module.tag = rep.tag;
```
Comment (gibfahn) [inline lib/lookup.js]: `tag`->`tags`

File: bin/citgm-all.js [hunk 8]
```diff
@@ -13,6 +13,7 @@ var reporter = require('../lib/reporter');
 var getLookup = require('../lib/lookup').get;
 var commonArgs = require('../lib/common-args');
 var isMatch = require('../lib/match-conditions');
+var checkTag = require('../lib/check-tag');
```
Comment (gibfahn) [inline bin/citgm-all.js]: We should probably call this `checkTags` (and change the file name).

File: bin/citgm-all.js [hunk 9]
```diff
@@ -30,6 +31,14 @@ yargs = commonArgs(yargs)
     alias: 'J',
     type: 'boolean',
     description: 'Auto detect number of cores to use to run tests in parallel'
+  })
+  .option('withTags', {
+    type: 'string',
+    description: 'define which tags from the lookup to run'
```
Comment (gibfahn) [inline bin/citgm-all.js]: `define` -> `Define`

File: bin/citgm-all.js [hunk 10]
```diff
@@ -30,6 +31,14 @@ yargs = commonArgs(yargs)
...
+  })
+  .option('excludeTags', {
+    type: 'string',
+    description: 'define which tags from the lookup to skip'
```
Comment (gibfahn) [inline bin/citgm-all.js]: `define` -> `Define`

File: lib/check-tag.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+});
+
+function checkTag(app, mod, name) {
+  if (mod.tag && mod.tag !== '' && typeof mod.tag === 'string') {
+    mod.tag = mod.tag.split(',');
+  }
+  if (app.withTags && typeof app.withTags === 'string') {
+    app.withTags = app.withTags.split(',');
+  }
+  if (app.excludeTags && typeof app.excludeTags === 'string') {
```
Comment (gibfahn) [inline lib/check-tag.js]: - In what situation are `app.withTags`, or `app.excludeTags` not undefined or of type `string`? `withTags` and `excludeTags` are defined as strings in the option parser. You could just drop this entirely and do this in the `checkMatches()` calls later (see later comment).

File: bin/citgm-all.js [hunk 12]
```diff
@@ -56,7 +65,9 @@ var options = {
   level: app.verbose,
   npmLevel: app.npmLoglevel,
   timeoutLength: app.timeout,
-  tmpDir: app.tmpDir
+  tmpDir: app.tmpDir,
+  withTags: app.withTags,
+  excludeTags: app.excludeTags
```
Comment (gibfahn) [inline bin/citgm-all.js]: 1. It seems a bit unnecessary to split `withTags` and `excludeTags` for each module, couldn't you just do it here?

```js
  withTags: app.withTags.split(','),
  excludeTags: app.excludeTags.split(',')
```

 2. Should we have some kind of logging at the start of the run case users use the wrong separator for tags, e.g. `citgm-all --withTags tag1 tag2` instead of `citgm-all --withTags tag1,tag2`. Something like:

```
Only running tests matching these 3 tags: tag1, tag2, tag3
Not running tests matching these 2 tags: tag4, tag5
```

to make it clear how many tags were found?

File: bin/citgm-all.js [hunk 13]
```diff
@@ -94,6 +105,10 @@ function runCitgm (mod, name, next) {
     return next();
   }
 
+  if (checkTag(app, mod, name)) {
```
Comment (gibfahn) [inline bin/citgm-all.js]: `checkTag` -> `checkTags`

File: lib/check-tag.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  } else if (app.excludeTags && mod.tag && checkMatch(mod.tag,
+        app.excludeTags)) {
+    log.info(name, 'skipped because of tag: ' + app.excludeTags);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+module.exports = checkTag;
```
Comment (gibfahn) [inline lib/check-tag.js]: `checkTag` -> `checkTags`

File: man/citgm-all.1 [hunk 15]
```diff
@@ -70,6 +70,12 @@ Run tests in parallel (automatically detect core count)
 .TP
 .BR \-\-tmpDir " " \fI<path>\fR
 Directory to test modules in
+.TP
+.BR \-\-withTags " " \fI<tag1,tag2>\fR
+Only test modules from the lookup that contain a matching tag field
+.TP
+.BR \-\-excludeTags " " \fI<tag1,tag2>\fR
+Specify which tags to skip from the lookup
```
Comment (gibfahn) [inline man/citgm-all.1]: These need to be updated if you change the equivalents in the README (a linter or even a test for this would be great).

File: test/fixtures/custom-lookup-tag.json [hunk 16]
```diff
@@ -0,0 +1,13 @@
+{
+  "omg-i-pass": {
+    "npm": true,
+    "tag": "tag1"
```
Comment (gibfahn) [inline test/fixtures/custom-lookup-tag.json]: `tag` -> `tags`

File: test/fixtures/custom-lookup-tag.json [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+  },
+  "omg-i-fail": {
+    "npm": true,
+    "tag": ["tag2", "tagNotUsed"]
```
Comment (gibfahn) [inline test/fixtures/custom-lookup-tag.json]: `tag` -> `tags`

File: lib/check-tag.js [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+  } else if (app.withTags && mod.tag && checkMatch(mod.tag, app.withTags)) {
+    log.info(name, 'the following tag was matched: ' + app.withTags);
+    return false;
+  } else if (app.excludeTags && mod.tag && checkMatch(mod.tag,
+        app.excludeTags)) {
+    log.info(name, 'skipped because of tag: ' + app.excludeTags);
+    return true;
+  } else {
+    return false;
+  }
```
Comment (gdams) [inline lib/check-tag.js]: we already are logging which tags are matched

File: bin/citgm-all.js (hunk 12 above)
Comment (gdams) [inline bin/citgm-all.js]: with regard to 1:
you can't use split there because the code will break if `withTags` or `excludeTags` is undefined.

Comment (gibfahn) [inline bin/citgm-all.js]: Okay, then you could do it in the line above right?

```js
  if (app.withTags) app.withTags.split(',');
  if (app.excludeTags) app.excludeTags.split(',');
```

You might have to use braces and more lines to appease the linter.

**Comment by gdams** [comment]
@MylesBorins @gibfahn updated PTAL

File: lib/check-tags.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
+var log = require('./out')({});
+
+function checkTags(app, mod, name) {
+  if (mod.tags && mod.tags !== '' && typeof mod.tags === 'string') {
```
Comment (gibfahn) [inline lib/check-tags.js]: As before, why is this needed? isn't mod.tags always an array or undefined?

File: lib/check-tags.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+      return (B.indexOf(tag) !== -1);
+    });
+  }
+
+  function checkMatches (A, B) {
+    var result = checkMatch(A, B);
+    if (result.length !== 0) {
+      return result;
+    } else {
+      return false;
```
Comment (gibfahn) [inline lib/check-tags.js]: Why return an array or a boolean? That seems needlessly confusing, you can just check the returned array's length (as before).

File: lib/check-tags.js [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+    }
+  }
+
+  if ((app.withTags && mod.tags && !checkMatches(mod.tags, app.withTags)) ||
+      (app.withTags && !mod.tags)) {
+    log.info(name, 'no matching tags found');
+    return true;
+  } else if (app.excludeTags && mod.tags && checkMatches(mod.tags,
+        app.excludeTags)) {
+    log.info(name, 'skipped because of tag: ' + checkMatches(mod.tags,
```
Comment (gibfahn) [inline lib/check-tags.js]: You're calling checkMatches twice, needlessly. Why not just use a variable (as before)?

File: lib/check-tags.js [hunk 22]
```diff
@@ -0,0 +1,41 @@
+var log = require('./out')({});
+
+function checkTags(app, mod, name) {
+  if (mod.tags && mod.tags !== '' && typeof mod.tags === 'string') {
+    mod.tags = mod.tags.split(',');
+  }
+
+  function checkMatch (A, B) {
+    return A.filter(function(tag) {
+      return (B.indexOf(tag) !== -1);
```
Comment (gibfahn) [inline lib/check-tags.js]: Is there a reason not to use an arrow function here? This seems like the perfect place for one (see earlier comment).

Also why have `checkMatches` calling `checkMatch`? This code checks for all matches, not just one. I don't see the need for two functions.

File: lib/check-tags.js [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    var result = checkMatch(A, B);
+    if (result.length !== 0) {
+      return result;
+    } else {
+      return false;
+    }
+  }
+
+  if ((app.withTags && mod.tags && !checkMatches(mod.tags, app.withTags)) ||
+      (app.withTags && !mod.tags)) {
```
Comment (gibfahn) [inline lib/check-tags.js]: This logic seems unnecessarily complex, and comments would be good, see my previous suggestion.

File: lib/check-tag.js (hunk 18 above)
Comment (gibfahn) [inline lib/check-tag.js]: You weren't, but I agree you are now.

File: README.md [hunk 24]
```diff
@@ -133,6 +135,7 @@ For syntax, see [lookup.json](./lib/lookup.json), the available attributes are:
 "sha": "<git-commit-sha>"    Test against a specific commit
 "envVar"                     Pass an environment variable before running
 "install": ["--param1", "--param2"] - Array of extra command line parameters passed to 'npm install'
+"tags": ["tag1", "tag2"]     Specify which tags apply to the module
```
Comment (gibfahn) [inline README.md]: Trailing whitespace

File: lib/check-tags.js [hunk 25]
```diff
@@ -0,0 +1,40 @@
+var log = require('./out')({});
+
+function checkTags(app, mod, name) {
+  if ((app.excludeTags && !mod.tags) || (!app.withTags && !app.excludeTags)) {
```
Comment (gibfahn) [inline lib/check-tags.js]: The empty array `[]` evaluates to true, so you'll need to use `app.withTags.length` and `app.excludeTags.length` in `if` conditions.

I'm not sure how this would pass the tests.

**Comment by gibfahn** [comment]
@MylesBorins does this LGTY?

**Comment by gdams** [comment]
CI: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-continuous-integration/17

**Comment by gdams** [comment]
CI: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-continuous-integration/18/

**Comment by gdams** [comment]
@nodejs/citgm can I get some reviews please?

**Comment by gdams** [comment]
@bzoz good spot! UPDATED

File: bin/citgm-all.js [hunk 26]
```diff
@@ -29,6 +30,14 @@ const yargs = commonArgs(require('yargs'))
     alias: 'J',
     type: 'boolean',
     description: 'Auto detect number of cores to use to run tests in parallel'
+  })
+  .option('withTags', {
+    type: 'string',
```
Comment (MylesBorins) [inline bin/citgm-all.js]: Is there a reason to not use the [array key](https://www.npmjs.com/package/yargs#arraykey) here

File: bin/citgm-all.js [hunk 27]
```diff
@@ -29,6 +30,14 @@ const yargs = commonArgs(require('yargs'))
...
+    description: 'Define which tags from the lookup to run'
+  })
+  .option('excludeTags', {
+    type: 'string',
```
Comment (MylesBorins) [inline bin/citgm-all.js]: see above

File: bin/citgm-all.js [hunk 28]
```diff
@@ -47,6 +56,17 @@ if (!app.su) {
   log.warn('root', 'Running as root! Use caution!');
 }
 
+if (app.withTags){
+  app.withTags = app.withTags.split(',');
```
Comment (MylesBorins) [inline bin/citgm-all.js]: this won't be necessary if we us the [array key](https://www.npmjs.com/package/yargs#arraykey)

File: bin/citgm-all.js [hunk 29]
```diff
@@ -47,6 +56,17 @@ if (!app.su) {
...
+  log.info('withTags', 'Only running tests matching these tags: '
+      + app.withTags);
+}
+if (app.excludeTags){
+  app.excludeTags = app.excludeTags.split(',');
```
Comment (MylesBorins) [inline bin/citgm-all.js]: see above

File: bin/citgm-all.js [hunk 30]
```diff
@@ -47,6 +56,17 @@ if (!app.su) {
   log.warn('root', 'Running as root! Use caution!');
 }
 
+if (app.withTags){
+  app.withTags = app.withTags.split(',');
+  log.info('withTags', 'Only running tests matching these tags: '
+      + app.withTags);
```
Comment (MylesBorins) [inline bin/citgm-all.js]: should we parse this out in a nicer way than simply printing the array?

File: bin/citgm-all.js [hunk 31]
```diff
@@ -47,6 +56,17 @@ if (!app.su) {
...
+}
+if (app.excludeTags){
+  app.excludeTags = app.excludeTags.split(',');
+  log.info('excludeTags', 'Not running tests matching these tags: '
+     + app.excludeTags);
```
Comment (MylesBorins) [inline bin/citgm-all.js]: see above

File: bin/citgm-all.js (hunk 12 above)
Comment (MylesBorins) [inline bin/citgm-all.js]: if you see my above comment we can use the yargs [array key](https://www.npmjs.com/package/yargs#arraykey) here to simplify things

File: bin/citgm-all.js [hunk 32]
```diff
@@ -97,6 +119,10 @@ function runCitgm (mod, name, next) {
   const runner = citgm.Tester(name, options);
   let bailed = false;
 
+  if (checkTags(app, mod, name)) {
+    return next(); // Skip this module
```
Comment (MylesBorins) [inline bin/citgm-all.js]: Are we generally putting the `return` and the call to `next` on the same line? Technically we don't want to return the result of the callback... but I'm not sure that it matters. Good to be explicit + consistent though. Maybe time for a new lint rule?

File: lib/check-tags.js [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+function checkTags(app, mod, name) {
+  // Returns true if the module should be skipped
+
+  if ((app.excludeTags && !mod.tags) || (!app.withTags && !app.excludeTags)) {
+    return false; // No checks to run.
+  } else if (app.withTags && !mod.tags) {
+    return true; // No tags for this module
+  }
+
+  if (typeof mod.tags === 'string') {
```
Comment (MylesBorins) [inline lib/check-tags.js]: If we use the array method mentioned above this should not be a necessary check, we will always get an array

File: lib/check-tags.js [hunk 34]
```diff
@@ -0,0 +1,43 @@
...
+    mod.tags = mod.tags.split(','); // If mod.tags is a single field
+  }
+
+  function checkMatches (A, B) {
+    return A.filter((tag) => B.indexOf(tag) !== -1);
+  }
+
+  let excludeTagMatches = app.excludeTags ?
```
Comment (MylesBorins) [inline lib/check-tags.js]: this check can be ignored if we make excludeTags default to an empty array

File: lib/check-tags.js [hunk 35]
```diff
@@ -0,0 +1,43 @@
...
+    log.info(name,
+        `skipped as these excludeTags matched: ${excludeTagMatches}`);
+    return true; // We matched an excludeTag.
+  } else if (app.excludeTags) {
+    log.info(`${name} will run as no excludeTags matched`);
+    return false; // We did not match an excludeTag.
+  }
+
+  let withTagMatches = app.withTags ?
+    checkMatches(mod.tags, app.withTags) : [];
```
Comment (MylesBorins) [inline lib/check-tags.js]: same as above

File: lib/check-tags.js [hunk 36]
```diff
@@ -0,0 +1,43 @@
...
+    return false; // No checks to run.
+  } else if (app.withTags && !mod.tags) {
+    return true; // No tags for this module
+  }
+
+  if (typeof mod.tags === 'string') {
+    mod.tags = mod.tags.split(','); // If mod.tags is a single field
+  }
+
+  function checkMatches (A, B) {
```
Comment (MylesBorins) [inline lib/check-tags.js]: this could be replaced with `_.instersection`

File: bin/citgm-all.js (hunk 26 above)
Comment (gdams) [inline bin/citgm-all.js]: changed

File: lib/check-tags.js (hunk 33 above)
Comment (gdams) [inline lib/check-tags.js]: this is to do with the lookup tags rather than than the args. So you could have `"tags": "test"` or `"tags": ["test1", "test2"]`

File: bin/citgm-all.js (hunk 32 above)
Comment (gdams) [inline bin/citgm-all.js]: we are generally at the moment, let's discuss elsewhere and find a linter rule once this is landed

File: bin/citgm-all.js (hunk 30 above)
Comment (gdams) [inline bin/citgm-all.js]: This just logs a comma separated output such as:
`Only running tests matching these tags: tag1, tag2`

File: bin/citgm-all.js (hunk 32 above)
Comment (gibfahn) [inline bin/citgm-all.js]: How else would you do it? If you do 

```js
return;
next();
```

then next will never get called right? Are you suggesting:

```js
next();
return;
```

**Comment by gdams** [comment]
okay so I have changed `withTags` to `includeTags`

File: bin/citgm-all.js (hunk 32 above)
Comment (targos) [inline bin/citgm-all.js]: It's also possible to write:

```js
return void next();
```

 I'm not saying we should do it, though.

File: bin/citgm-all.js [hunk 37]
```diff
@@ -97,6 +117,10 @@ function runCitgm (mod, name, next) {
   const runner = citgm.Tester(name, options);
   let bailed = false;
 
+  if (checkTags(app, mod, name)) {
```
Comment (targos) [inline bin/citgm-all.js]: I guess this can be moved up before the runner is instanciated?

File: lib/check-tags.js [hunk 38]
```diff
@@ -0,0 +1,41 @@
+'use strict';
+const log = require('./out')({});
```
Comment (targos) [inline lib/check-tags.js]: I think it's the first file outside of `bin` that directly requires this file. Shouldn't it receive the context and use its `data` event?
In that case, my previous comment becomes obsolete.

File: lib/check-tags.js [hunk 39]
```diff
@@ -0,0 +1,41 @@
...
+  // Returns true if the module should be skipped
+
+  if ((app.excludeTags && !mod.tags) || (!app.includeTags && !app.excludeTags)) {
+    return false; // No checks to run.
+  } else if (app.includeTags && !mod.tags) {
+    return true; // No tags for this module
+  }
+
+  if (typeof mod.tags === 'string') {
+    mod.tags = mod.tags.split(','); // If mod.tags is a single field
```
Comment (targos) [inline lib/check-tags.js]: I think `mod.tags = [mod.tags]` is enough here.
If it is intentional to accept multiple tags in a string, this should be documented and tested.

File: lib/check-tags.js [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+
+  let excludeTagMatches = _.intersection(app.excludeTags, mod.tags);
+
+  if (excludeTagMatches.length) {
+    log.info(name,
+        `skipped as these excludeTags matched: ${excludeTagMatches}`);
+    return true; // We matched an excludeTag.
+  } else if (app.excludeTags) {
+    log.info(`${name} will run as no excludeTags matched`);
+    return false; // We did not match an excludeTag.
```
Comment (targos) [inline lib/check-tags.js]: So, you couldn't combine `includeTags` and `excludeTags` ? Seems like an unnecessary limitation.

File: lib/check-tags.js [hunk 41]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  let includeTagMatches = _.intersection(app.includeTags, mod.tags);
+
+  if (includeTagMatches.length) {
+    log.info(name, `will run as these includeTags matched: ${includeTagMatches}`);
+    return false; // We matched a includeTag.
+  } else {
+        console.error('here')
```
Comment (targos) [inline lib/check-tags.js]: remaining debug code? ;)

File: lib/check-tags.js [hunk 42]
```diff
@@ -0,0 +1,41 @@
...
+    log.info(name, 'skipped as no includeTags matched');
+    return true; // We didn't match any withTags.
```
Comment (targos) [inline lib/check-tags.js]: "We did not match any includeTag." (for consistency with the excludeTag comment

File: man/citgm-all.1 [hunk 43]
```diff
@@ -70,6 +70,12 @@ Run tests in parallel (automatically detect core count)
 .TP
 .BR \-\-tmpDir " " \fI<path>\fR
 Directory to test modules in
+.TP
+.BR \-\-includeTags " " \fI<tag1,tag2>\fR
```
Comment (targos) [inline man/citgm-all.1]: Is it the correct syntax? The parameters are not supposed to be comma-separated.

File: lib/check-tags.js (hunk 41 above)
Comment (gdams) [inline lib/check-tags.js]: removed

File: man/citgm-all.1 (hunk 43 above)
Comment (gdams) [inline man/citgm-all.1]: good spot

File: lib/check-tags.js (hunk 42 above)
Comment (gdams) [inline lib/check-tags.js]: updated

File: lib/check-tags.js (hunk 38 above)
Comment (gdams) [inline lib/check-tags.js]: updated

File: lib/check-tags.js (hunk 40 above)
Comment (gibfahn) [inline lib/check-tags.js]: I guess it makes sense to allow `excludeTags` and `includeTags`.

If you have:

```json
"both": {
  "tags": ["a", "b"]
},
"aOnly": {
  "tags": ["a"]
},
"bOnly": {
  "tags": ["b"]
},
"neither": {
  "tags": []
}
```

| arguments | Modules run |
|--- | --- |
| `` | `both badOnly goodOnly neither` |
| `--includeTags=a` \| `--excludeTags=b` | `both aOnly` |
| `--excludeTags=a` \| `--includeTags=b` | `neither bOnly` |
| `--includeTags=a --excludeTags=b` | `bOnly` |

Comment (targos) [inline lib/check-tags.js]: @gibfahn I would have a different expectation with the same list:


| arguments | Modules run |
|--- | --- |
| `` | `both badOnly goodOnly neither` |
| `--includeTags=a` | `both aOnly` |
| `--excludeTags=a` | `bOnly neither` |
| `--includeTags=b` | `both bOnly` |
| `--excludeTags=b` | `aOnly neither` |
| `--includeTags=a --excludeTags=b` | `aOnly` |
| `--includeTags=b --excludeTags=a` | `bOnly` |

Comment (gibfahn) [inline lib/check-tags.js]: @targos Yeah sorry, I changed the example half way through editing it and messed it up.

File: bin/citgm-all.js [hunk 44]
```diff
@@ -47,6 +56,15 @@ if (!app.su) {
   log.warn('root', 'Running as root! Use caution!');
 }
 
+if (app.includeTags){
```
Comment (gdams) [inline bin/citgm-all.js]: it is by default undefined

File: lib/check-tags.js (hunk 40 above)
Comment (gibfahn) [inline lib/check-tags.js]: @gdams how about this:

```js
'use strict';
const _ = require('lodash');

function checkTags(app, mod, name, log) {
  // Returns true if the module should be skipped.

  if ((app.excludeTags && !mod.tags) ||
      (!app.includeTags && !app.excludeTags)) {
    return false; // No checks to run.
  } else if (app.includeTags && !mod.tags) {
    return true; // No tags for this module.
  }

  if (typeof mod.tags === 'string') {
    mod.tags = [mod.tags]; // If mod.tags is a single string, convert to array.
  }

  let excludeTagMatches = _.intersection(app.excludeTags, mod.tags);

  if (excludeTagMatches.length) {
    log.info(name,
        `skipped as these excludeTags matched: ${excludeTagMatches}`);
    return true; // We matched an excludeTag.
  }

  let includeTagMatches = _.intersection(app.includeTags, mod.tags);

  if (includeTagMatches.length) {
    log.info(name,
        `will run as these includeTags matched: ${includeTagMatches}`);
    return false; // We matched an includeTag.
  }

  if (app.includeTags) {
    log.info(`${name} skipped  as no includeTags matched`);
    return false; // We did not match an includeTag.
  } else {
    log.info(`${name} will run as no excludeTags matched`);
    return true; // We did not match an excludeTag.
  }


}

module.exports = checkTags;
```

**Comment by gdams** [comment]
@targos @gibfahn UPDATED PTAL, this also includes tests for every use case discussed above

File: lib/check-tags.js [hunk 45]
```diff
@@ -0,0 +1,44 @@
...
+  let includeTagMatches = _.intersection(app.includeTags, mod.tags);
+
+  if (includeTagMatches.length) {
+    log.info(name,
+        `will run as these includeTags matched: ${includeTagMatches}`);
+    return false; // We matched an includeTag.
+  }
+
+  if (app.includeTags) {
+    log.info(`${name} skipped  as no includeTags matched`);
```
Comment (targos) [inline lib/check-tags.js]: nit: there are 2 spaces between "skipped" and "as"

File: test/test-check-tags.js [hunk 46]
```diff
@@ -0,0 +1,297 @@
...
+  };
+  const mod = {
+    tags: ['a', 'b']
+  };
+  t.plan(1);
+  const result = checkTags(app, mod, 'test', log);
+  t.true(result, 'should return true');
+});
+
+test('test excludeTags and matching tag', function (t) {
```
Comment (targos) [inline test/test-check-tags.js]: nit: no matching tag

File: test/test-check-tags.js [hunk 47]
```diff
@@ -0,0 +1,297 @@
...
+  };
+  const mod = {
+    tags: 'a'
+  };
+  t.plan(1);
+  const result = checkTags(app, mod, 'test', log);
+  t.false(result, 'should return false');
+});
+
+test('test excludeTags and no matching tag', function (t) {
```
Comment (targos) [inline test/test-check-tags.js]: nit: matching tag

File: test/test-check-tags.js (hunk 46 above)
Comment (gdams) [inline test/test-check-tags.js]: updated

File: lib/check-tags.js [hunk 48]
```diff
@@ -0,0 +1,44 @@
...
+    return false; // No checks to run.
+  } else if (app.includeTags && !app.excludeTags && !mod.tags) {
+    return true; // No tags for this module.
+  }
+
+  if (typeof mod.tags === 'string') {
+    mod.tags = [mod.tags]; // If mod.tags is a single string, convert to array.
+  }
+
+  let excludeTagMatches = _.intersection(app.excludeTags, mod.tags);
```
Comment (MylesBorins) [inline lib/check-tags.js]: app.excludeTags || []

File: bin/citgm-all.js (hunk 44 above)
Comment (MylesBorins) [inline bin/citgm-all.js]: If it is by default undefined then I think we need to do some safety checks inside of check-tags to ensure that we are not checking for .length on undefined

File: lib/check-tags.js (hunk 48 above)
Comment (gdams) [inline lib/check-tags.js]: updated

File: bin/citgm-all.js [hunk 49]
```diff
@@ -55,7 +73,9 @@ const options = {
   level: app.verbose,
   npmLevel: app.npmLoglevel,
   timeoutLength: app.timeout,
-  tmpDir: app.tmpDir
+  tmpDir: app.tmpDir,
+  includeTags: app.includeTags,
```
Comment (MylesBorins) [inline bin/citgm-all.js]: ohhhh we could do it here... includeTags: app.includeTags || []

Comment (gibfahn) [inline bin/citgm-all.js]: If you do this I think you'll have to change all the `if(app.includeTags)` conditions to `if(app.includeTags.length)`.

File: lib/check-tags.js [hunk 50]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+const _ = require('lodash');
+
+function checkTags(options, mod, name, log) {
+  // Returns true if the module should be skipped.
+
+  if ((options.excludeTags.length && !options.includeTags.length && !mod.tags)
+    || (!options.includeTags.length && !options.excludeTags.length)) {
+    return false; // No checks to run.
+  } else if (options.includeTags.length && !options.excludeTags.length
```
Comment (MylesBorins) [inline lib/check-tags.js]: do we need to include `!options.excludeTags.length` here? If there are include tags, but no exclude tags, and there are no mod tags, then we definitely want to skip

File: lib/check-tags.js [hunk 51]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  if (options.includeTags.length) {
+    log.info(`${name} skipped as no includeTags matched`);
+    return true; // We did not match an includeTag.
+  } else {
+    log.info(`${name} will run as no excludeTags matched`);
+    return false; // We did not match an excludeTag.
+  }
+
```
Comment (MylesBorins) [inline lib/check-tags.js]: are we 100% there are no edges where we end up here with no return?

Comment (gdams) [inline lib/check-tags.js]: I have got tests for every outcome using `--includeTags` and `--excludeTags`

**Comment by gdams** [comment]
all tests passed apart from OSX which we are having issues with
